src/miniscript/mod.rs

```diff
@@ -26,114 +26,127 @@
 
 #[cfg(feature = "serde")]
 use serde::{de, ser};
+use std::marker::PhantomData;
 use std::{fmt, str};
 
 use bitcoin;
 use bitcoin::blockdata::script;
 
+pub use self::context::Legacy;
+pub use self::context::Segwitv0;
+
 pub mod astelem;
+pub(crate) mod context;
 pub mod decode;
 pub mod lex;
 pub mod satisfy;
 pub mod types;
 
 use self::lex::{lex, TokenIter};
 use self::types::Property;
+pub use miniscript::context::ScriptContext;
+use miniscript::decode::Terminal;
 use miniscript::types::extra_props::ExtData;
 use miniscript::types::Type;
+
 use std::cmp;
 use std::sync::Arc;
 use MiniscriptKey;
 use {expression, Error, ToPublicKey};
 
 /// Top-level script AST type
 #[derive(Clone, Hash)]
-pub struct Miniscript<Pk: MiniscriptKey> {
+pub struct Miniscript<Pk: MiniscriptKey, Ctx: ScriptContext> {
     ///A node in the Abstract Syntax Tree(
-    pub node: decode::Terminal<Pk>,
+    pub node: Terminal<Pk, Ctx>,
     ///The correctness and malleability type information for the AST node
     pub ty: types::Type,
     ///Additional information helpful for extra analysis.
     pub ext: types::extra_props::ExtData,
+    /// Context PhantomData
+    pub phantom: PhantomData<Ctx>,
```

Comment:
I think we shouldn't make this public.

This has the weird effect of making it impossible to construct a `Miniscript` without going through the constructor, but I'm not sure we ever wanted that. The other fields are public to make introspection possible, but ideally we don't want the user to be modifying these things.